File: programmers/Level2/solution/_18_피로도/Solution.java
package programmers.Level2.solution._18_피로도;

/**
 * https://school.programmers.co.kr/learn/courses/30/lessons/87946?language=java
 */
public class Solution {

    public static void main(String[] args) {

        int k = 80;
        int[][] dungeons = {
                { 80, 20 }, { 50, 40 }, { 30, 10 }
        };

        Solution solution = new Solution();
        int result = solution.solution(k, dungeons);

        System.out.println(result);
    }

    public static boolean[] visited;
    public static int result = 0;

    public int solution(int k, int[][] dungeons) {
        visited = new boolean[dungeons.length];
        dfs(k, dungeons, 0);
        return result;
    }

    public void dfs(int k, int[][] dungeons, int depth) {
        for (int i = 0; i < dungeons.length; i++) {
            if (dungeons[i][0] <= k && !visited[i]) {
                visited[i] = true;
                dfs(k - dungeons[i][1], dungeons, depth + 1);
                visited[i] = false;
            }
        }
        result = Math.max(result, depth);
    }
}

File: programmers/Level2/solution/_18_피로도/Solution.kt
/*
 * https://school.programmers.co.kr/learn/courses/30/lessons/87946?language=kotlin
 */
fun main() {
    var k = 80
    val dungeons: Array<IntArray> = arrayOf(
        intArrayOf(80, 20),
        intArrayOf(50, 40),
        intArrayOf(30, 10)
    )

    var result = solution(k, dungeons)
    print("result: " + result)
}

var visited = BooleanArray(0)
var answer = 0

fun solution(k: Int, dungeons: Array<IntArray>): Int {
    visited = BooleanArray(dungeons.size)
    dfs(k, dungeons, 0)
    return answer
}

fun dfs(k: Int, dungeons: Array<IntArray>, depth: Int) {
    for (i in dungeons.indices) {
        if (dungeons[i][0] <= k && !visited[i]) {
            visited[i] = true
            dfs(k - dungeons[i][1], dungeons, depth + 1)
            visited[i] = false
        }
    }
    answer = Math.max(answer, depth)
}
